cachi2/core/package_managers/yarn_classic/resolver.py

```diff
@@ -47,26 +53,99 @@ class _RelpathMixin(BaseModel):
 class RegistryPackage(_BasePackage, _UrlMixin):
     """A Yarn 1.x package from the registry."""
 
+    @property
+    def purl(self) -> str:
+        """Return package URL."""
+        qualifiers = {}
+
+        if self.url != DEFAULT_NPM_REGISTRY:
+            qualifiers = {"repository_url": ALTERNATIVE_NPM_REGISTRY}
+
+        if self.integrity:
+            checksum = ChecksumInfo.from_sri(self.integrity)
+            qualifiers["checksum"] = str(checksum)
+
+        return PackageURL(
+            type="npm",
+            name=self.name,
+            version=self.version,
+            qualifiers=qualifiers,
+        ).to_string()
+
 
 class GitPackage(_BasePackage, _UrlMixin):
     """A Yarn 1.x package from a git repo."""
 
+    @property
+    def purl(self) -> str:
+        """Return package URL."""
+        parsed_url = urlparse(self.url)
+        repo_id = RepoID(origin_url=self.url, commit_id=parsed_url.fragment)
+        qualifiers = {"vcs_url": repo_id.as_vcs_url_qualifier()}
+        return PackageURL(
+            type="npm",
+            name=self.name,
+            version=self.version,
+            qualifiers=qualifiers,
+        ).to_string()
+
 
 class UrlPackage(_BasePackage, _UrlMixin):
     """A Yarn 1.x package from a http/https URL."""
 
+    @property
+    def purl(self) -> str:
+        """Return package URL."""
+        qualifiers = {"download_url": self.url}
+        return PackageURL(
+            type="npm",
+            name=self.name,
```

Comment:
For some of the non-registry packages, we may want to read their respective package.json files from the cached tarballs to get their true names. As a user, I can call these non-registry dependencies whatever I want in my project's package.json and it could be completely different from what it really is:
```
"dependencies": {
  "not-fecha": "https://github.com/taylorhakes/fecha.git"
}
```

Reply:
Does the filename in the cache come from package.json or the URL ?

Reply:
I believe it is the filename from the resolved URL

Reply:
That's a tricky situation. Theoritacily, we could have three potential names:

- the name declared in the project's package.json
- the actual package name from the dependency's package.json
- the repository name from the Git URL

Reply:
Is this only relevant for `GitPackage` and `UrlPackage` ?

Reply:
We actually have [precedent](https://github.com/containerbuildsystem/cachi2/blob/main/cachi2/core/package_managers/yarn/resolver.py#L356) for using the name present in the `package.json` file in the yarn v3 implementation. 

We don't support git deps in the Yarn v3 backend, but I think we should use the same approach for both file and git deps.